apiserver/facades/client/secrets/secrets.go

```diff
@@ -406,16 +407,48 @@ func (s *SecretsAPI) updateSecret(backend provider.SecretsBackend, arg params.Up
 			}
 		}
 	}
-	_, err = s.secretsState.UpdateSecret(uri, fromUpsertParams(arg.UpsertSecretArg))
-	return errors.Trace(err)
+	md, err = s.secretsState.UpdateSecret(uri, fromUpsertParams(arg.AutoPrune, arg.UpsertSecretArg))
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if md.AutoPrune {
+		// If the secret was updated, we need to delete the old unused secret revisions.
+		revsToDelete, err := s.secretsState.ListUnusedSecretRevisions(uri)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		pruneArg := params.DeleteSecretArg{URI: md.URI.String()}
+		for _, rev := range revsToDelete {
+			if rev == md.LatestRevision {
+				// We don't want to delete the latest revision.
+				continue
+			}
+			pruneArg.Revisions = append(pruneArg.Revisions, rev)
+		}
+		if len(pruneArg.Revisions) == 0 {
+			return nil
+		}
+		pruneResult, err := s.RemoveSecrets(params.DeleteSecretArgs{Args: []params.DeleteSecretArg{pruneArg}})
+		if err != nil {
+			// We don't want to fail the update if we can't prune the unused secret revisions because they will be picked up later
+			// when the secret has any new obsolute revisions.
+			logger.Warningf("failed to prune unused secret revisions for %q: %v", uri, err)
```

Comment:
We should surface this to the user - we could return a typed "failed to prune" error which the caller can choose to just print out and ignore etc.

Reply:
Ideally, we should do these prune in the secretPruner worker.
I created a card `JUJU-4723`, we will do it in the following PRs.